File: Cobluesky/Programmers - 콜라츠 추측.js
//https://school.programmers.co.kr/learn/courses/30/lessons/12943
function solution(num) {
    var answer = 0;

    while (num !== 1) // 숫자가 1인가?
    {
        if (answer === 500) // count가 500인가?
        {
            return -1;
        }

        if (num % 2 === 0) // 짝수인가?
        {
            num = (num / 2);
            answer++;
        }

        else // 홀수인가?
        {
            num = (num * 3) + 1;
            answer++;
        }
    }
    return answer;
}

// 대체로 for보다는 while이 빠릅니다.
// for each를 쓸 수 있으면 for each가 무조건 빠릅니다만 for each는 object만 사용이 가능합니다.
// 제 생각에는 시간복잡도가 괜찮게 나온 것 같습니다.

File: sollyj/programmers/Programmers_콜라츠추측.java
package sollyj.programmers;

public class Programmers_콜라츠추측 {
	public static void main(String[] args) {
		System.out.println(solution(626331));
	}

	private static int solution(int num) {
		int answer = 0;
		long n = num;    // 입력값이 크면 연산을 하다가 값이 커져 int 오버플로우 현상이 나타난다. 그래서 long형으로 임의로 바꿔주였다.

		while (answer <= 500) {
			if (n == 1) {
				break;
			}

			if (n % 2 == 0) {
				n /= 2;
			} else {
				n *= 3;
				n += 1;
			}

			answer++;
		}

		if (answer > 500)
			answer = -1;

		return answer;
	}
}
